pkg/validate/container_service.go

```diff
@@ -1,10 +1,258 @@
+	}
+
+	if c.Properties == nil {
+		errs = append(errs, fmt.Errorf("properties cannot be nil"))
+		return
+	}
+
+	errs = append(errs, validateProperties(c.Properties)...)
+	return
+}
+
+func validateProperties(p *api.Properties) (errs []error) {
+	errs = append(errs, validateProvisioningState(p.ProvisioningState)...)
+	errs = append(errs, validateOrchestratorProfile(p.OrchestratorProfile)...)
+	errs = append(errs, validateMasterProfile(p.MasterProfile)...)
+	errs = append(errs, validateAgentPoolProfiles(p.AgentPoolProfiles)...)
+	errs = append(errs, validateServicePrincipalProfile(p.ServicePrincipalProfile)...)
+	return
+}
+
+func validateServicePrincipalProfile(spp *api.ServicePrincipalProfile) (errs []error) {
+	if spp == nil {
+		errs = append(errs, fmt.Errorf("servicePrincipalProfile cannot be nil"))
+		return
+	}
+	if spp.ClientID == "" {
+		errs = append(errs, fmt.Errorf("invalid properties.servicePrincipalProfile.clientId %q", spp.ClientID))
+	}
+
+	if spp.Secret == "" {
+		errs = append(errs, fmt.Errorf("invalid properties.servicePrincipalProfile.secret %q", spp.Secret))
+	}
+
+	return
+}
+
+func validateAgentPoolProfiles(apps []*api.AgentPoolProfile) (errs []error) {
+	appmap := map[string]*api.AgentPoolProfile{}
+
+	for i, app := range apps {
+		// TODO why is this a pointer?
+		if app == nil {
+			continue
+		}
+
+		if _, found := validAgentPoolProfileNames[app.Name]; !found {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q]", app.Name))
+		}
+
+		if _, found := appmap[app.Name]; found {
+			errs = append(errs, fmt.Errorf("duplicate properties.agentPoolProfiles %q", app.Name))
+		}
+		appmap[app.Name] = app
+
+		if i > 0 && app.VnetSubnetID != apps[i-1].VnetSubnetID {
+			errs = append(errs, fmt.Errorf("duplicate properties.agentPoolProfiles.vnetSubnetID %q", app.VnetSubnetID))
+		}
+
+		errs = append(errs, validateAgentPoolProfile(app)...)
+	}
+
+	for name := range validAgentPoolProfileNames {
+		if _, found := appmap[name]; !found {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q]", name))
+		}
+	}
+
+	return
+}
+
+func validateAgentPoolProfile(app *api.AgentPoolProfile) (errs []error) {
+	if app.Name != string(app.Role) {
+		errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].name %q", app.Name, app.Name))
+	}
+
+	switch app.Role {
+	case api.AgentPoolProfileRoleCompute,
+		api.AgentPoolProfileRoleInfra:
+		if app.Count < 1 || app.Count > 100 {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].count %q", app.Name, app.Count))
+		}
+
+	case api.AgentPoolProfileRoleMaster:
+		if app.Count != 3 {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].count %q", app.Name, app.Count))
+		}
+
+	default:
+		errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].role %q", app.Name, app.Role))
+	}
+
+	switch app.VMSize {
+	case "Standard_D2s_v3",
```

Comment:
Are these the only two VM size OSA will allow? And why?

Reply:
These are the sizes we know are min required, we should allow anything bigger though.  Not really relevant to this code move (wasn't added in this PR, just moved).  We can expand if necessary.